ENH: Integrating LAPACK "expert" routines with conditioning warnings for linalg.solve_ family

ilayn opened the pull request:
**This is a rework of #6693** 

As briefly discussed in #6634, in this PR the ?GESVX routine of LAPACK is exposed for generic use. They are the so-called eXpert versions of the already existing ?GESV routine family which are used in `sp.linalg.solve` (also ?SYSVX, ?HESVX, ?POSVX, ?GECON, ?POCON) are also exposed).

The use case for these routines are many-fold 
- The most important case is to finally change the silent singular matrix handling. With the condition number and error estimates, a warning/error/msg can be emitted (needs-decision).
- Ill-conditioned matrices can be checked before inversion steps and nonsensical results can be avoided
- cond and rcond functions can now use estimates for comparison
- the lu factorization comes with a bonus (as is with ?GESV too but the API throws it away currently)
- The numerical accuracy is improved thanks to the balancing and iterative refinement (Lapack prefers the term `equilibration` for this particular instance). 
- The right inversion is currently a bit annoying, that is to say if we would like to compute, say, `A B⁻¹ `, then the user has to explicitly transpose the explicitly transposed equation to avoid using an explicit  `inv` whose use should be discouraged anyways: 
             ` x = scipy.linalg.solve(B.T,A.T).T` 
Since expert drivers come with a `trans` switch that can internally handle whether to solve the transposed or the regular equation, these routines avoid the `A.T` off-the-shelf. 

However, in the future, together with a low-level implicit transpose operation this can be handled rather more elegantly. Hence it can act like a forward or backward slash operator.

There will be a a few phases for this PR to come to fruition and I would really appreciate all the feedback from the wizards.
- [x] Make sure the f2py interface is correct. 
- [x] Make sure that the input-output arguments are handled correctly.
- [x] Make sure that the varying number of I/O arguments are not exposed to the user and the decision tree should not lead to confusion. In other words, if out of inexperience a user selects two conflicting keyword arguments (as it is extremely easy to fall into those traps) the behavior should be unambiguous override or err out directly. 
- [x] Make sure that the tests cover sufficient examples
- [x] Decide on the numerical bound formula to accept numerical singularity
- [x]  Decide on where this functionality goes into, existing `sp.linalg.solve` as an addition or a standalone solver and so on.

-----------------------

Example use-case, with warning

```
import numpy as np
import scipy.linalg as la
A = np.array([[1,2,3],[4,5,6],[7,8,9]],dtype=float)
b = np.array([[15,15,15]]).T
la.solve_x(A,b)

/usr/local/lib/python3.5/dist-packages/scipy/linalg/basic.py:224: 
RuntimeWarning: Ill-conditioned matrix detected.
Result is not guaranteed to be accurate.
Reciprocal condition number: 1.5419764230904951e-18

array([[-39.],
       [ 63.],
       [-24.]])
```

----------------------------------

Here are the speed comparison calling directly `sp.linalg.lapack.dgesvx(A,B)` and `sp.linalg.lapack.dgesv(A,B)` with 
```
A = np.random.rand(n, n)
B = np.random.rand(n, 3)
```

horizontal axis is `n` and vertical axis is seconds in the first one. In the second it's just the ratio of two curves. The numbers are averages of 25 runs for each n. (The blue one is the current DGESV and the other one is the proposed DGESVX)

![1](https://cloud.githubusercontent.com/assets/1303842/20037654/540d643e-a424-11e6-9d6b-860c47e7b5d6.png)
![2](https://cloud.githubusercontent.com/assets/1303842/20037655/56a1215e-a424-11e6-8569-b629fcd4cada.png)

person142 commented:
Finally got around to looking at this...

Since we get better accuracy and warnings about ill conditioned matrices, I'm +1 for using `dgesvx` in `solve`. (Might be good to come up with an example that highlights the accuracy gains.) I like the keyword that lets the user specify the structure of the matrix, though I'm not sure how to insert it into `solve` in a backwards-compatible way.

ilayn commented:
My vote, for what is worth, goes for deprecating the `sym_pos` keyword as it should probably have never been there in the first place in such a wide-use generic function. In the meantime `sym_pos` can point to `assume_a='pos'` internally during the deprecation period and eventually removed. This would allow future additions rather convenient as the API would stay fixed but possible options can be increased. 

`lower` is picked up as is and it would be still backwards-compatible.

The `debug` keyword is there but I don't know for which reason it is there since it only prints the overwrite keywords.

person142 commented:
The unfortunate thing is that this would require putting `assume_a` last so as to not break existing code, which is not where it should be...

Might be cleaner to have `solve_x`, deprecate `solve`, and then make `solve_x` become `solve` in 1.0?

ilayn commented:
Oh for that, I think I am not qualified to make a decision since you guys have been dealing with all kinds of user reported issues. I don't feel like I have the necessary experience about the releases.

rgommers commented:
> Might be cleaner to have ``solve_x``, deprecate ``solve``, and then make ``solve_x`` become ``solve`` in 1.0?

Can't really do that, both because the time for deprecation would be much too short and (more importantly) because we don't want to change API/behavior like that. Adding a new function and deprecating the old one is fine, but then renaming new -> old is not.

person142 commented:
So something like changing the signature to `solve(a, b, sym_pos=False, ..., assume_a='gen')`, deprecating `sym_pos`, and then eventually changing it to `solve(a, b, assume_a='gen')` would be out too?

ilayn commented:
I would like to ping this for a gentle reminder. Could you please maybe add `needs-decision` tag to draw attention of others too? 

If we decide to continue with this I'll start adding more tests. Otherwise, well, tough luck! 😁

rgommers commented:
> So something like changing the signature to `solve(a, b, sym_pos=False, ..., assume_a='gen')`, deprecating `sym_pos`, and then eventually changing it to `solve(a, b, assume_a='gen')` would be out too?

I'm afraid so.

This PR has signature:

    def solve_x(a, b, assume_a='gen', transposed=False, lower=False,
                overwrite_a=False, overwrite_b=False, check_finite=True):

The current signature for ``solve`` is:

    def solve(a, b, sym_pos=False, lower=False, overwrite_a=False,
              overwrite_b=False, debug=False, check_finite=True):

So we have a ``debug`` keyword now that's ugly, and a ``sym_pos`` one that's suboptimal but with functionality that will still be useful (because it's equivalent to ``assume_a='pos'``. 

The difference between the optimal signature and 

    def solve(a, b, sym_pos=False, lower=False, overwrite_a=False,
              overwrite_b=False, debug=False, check_finite=True, assume_a='gen'):

is imho not worth breaking backwards compat for or deprecating ``solve`` in favor of a new name.

Any other opinions?

rgommers commented:
> If we decide to continue with this I'll start adding more tests

@ilayn I think we're agreed on that this is a good addition, so go ahead with adding tests I'd say. The signature we'll be able to work out.

person142 commented:
Just to second what @rgommers already said--definitely +1 on the addition. Kept meaning to respond but then got caught up thinking about the API...

ilayn commented:
OK, clear then. Could you also please consider whether the name is a proper one? `solve_x` sounds a bit cheesy and science fiction-ish... (I'm really trying hard not to make a Sci-Fy pun).

person142 commented:
> Could you also please consider whether the name is a proper one?

I think we're in favor of replacing `solve` with `solve_x` by adding the `assume_a` keyword to the end of the signature of solve? @rgommers?

rgommers commented:
> I think we're in favor of replacing solve with solve_x by adding the assume_a keyword to the end of the signature of solve? @rgommers?

Agreed. Let's consider that decided unless someone disagrees within the next 24 hours.

ilayn commented:
I've found another discrepancy in the behaviors of both and a bug. 
```
sp.linalg.solve(1,[1, 2, 3])          # it gives *matrix not square error*. 
sp.linalg.solve([[1]], [1, 2, 3]      # then it errs with size mismatch. 
sp.linalg.solve(np.eye(3), [1,2,3])   # is valid and result still 1D array. should have complained
```
In the last case, the result can still survive the matrix multiplications and similar ops thanks to the fact that `.dot()` treats 1D arrays as inner products. However this fails when you assign the result to a slice with broadcasting for example:
```
x_1d = np.array([1,2,3,4])
x_2d = np.array([[1,2,3,4]]).T
zzz = np.zeros((4,3))
zzz[:, [1,2]] = x_2d # This works with broadcasting
zzz[:, [1,2]] = x_1d # This gives ValueError shape mismatch:
```

So should we fix the scalar bug and pretend this never happened or should we return a 2D array result compatible with shape of a? 

For what its worth this behavior is not optimal and I'm not sure whether this is a good thing from the 1D arrays are numpy's way of working point of view. As a person who constantly fights with 1D array mayhem not fitting anywhere, I think this is not a good behavior but since nobody seems to be bothered with it, I don't have a strong stance.

ilayn commented:
All green except some thread safety test and I think I've addressed all issues as far as I understand. So I think this is ready for review. 

I will try to see what that failing test is. Didn't see why it fails yet.

ilayn commented:
Is it possible that I have to add `threadsafe` (in one of the f2py sigs) somewhere to make it pass?

ilayn commented:
Is there anyone who can replicate this error? It doesn't happen on my virtual machines unfortunately.

Eric89GXL commented:
Have you tried spinning up a Travis-like virtualbox image using Ubuntu 12.04. For other details you can look here:

https://docs.travis-ci.com/user/ci-environment/

And then follow the Travis build setup. It's a bit annoying to try, but it might be worth it if you want to replicate the error.

ilayn commented:
@Eric89GXL Indeed, that's what I have tried to do on a VM but still no luck yet. I am not sure what the test is supposed to be testing. I'll just add `threadsafe` to everything and see whether the problem goes away.

ilayn commented:
@pv I see that you have committed also to this file, could you please have a look at it when you have time?

ilayn commented:
OK sporadic addition of `threadsafe` keyword in the LAPACK routines solved the issue.

ilayn commented:
Pinging for a gentle reminder about this

pv commented on scipy/linalg/basic.py:
```diff
@@ -75,34 +101,117 @@ def solve(a, b, sym_pos=False, lower=False, overwrite_a=False,
+                                       )
+    elif _structure == 'sym':
+        sysvx, sysvx_lw = get_lapack_funcs(('sysvx', 'sysvx_lwork'), (a1, b1))
+        lwork, _ = sysvx_lw(n, lower)
```
This, and the others, likely should use `_compute_lwork` --- for reasons see _compute_lwork

pv commented on scipy/linalg/tests/test_basic.py:
```diff
@@ -602,6 +604,97 @@ def test_check_finite(self):
+        a = 1
+        b = [1, 2, 3]
+        x = solve(a, b)
+        assert_array_almost_equal(x.ravel(), b)
```
Should this also explicitly check `x.shape`?

ilayn replied:
It should. I'll add another test for it.

pv commented:
I think otherwise this looks fine, and we want it to work like this, so removing the needs-decision flag.

pv commented on scipy/linalg/basic.py:
```diff
@@ -23,21 +23,43 @@
 # Linear equations
 def solve(a, b, sym_pos=False, lower=False, overwrite_a=False,
-          overwrite_b=False, debug=False, check_finite=True):
+          overwrite_b=False, check_finite=True, assume_a='gen',
```
The `debug=` should be retained.

pv replied:
I think better separate the backwards incompatible change from this PR, then we don't have to think about it now.

ilayn commented:
@pv Do you have any opinion why `threadsafe` actually mattered here ?

pv commented:
My guess is that adding threadsafe did not actually matter, it just re-ran the test. The test is timing based with 10ms delays, so it's possible to get sporadic failure with bad luck, which I think occurred here.

On the other hand, the LAPACK routines should be threadsafe, so there's no issue adding the statement --- and it can be beneficial. I don't recall exactly what it does, but maybe it releases GIL.

ilayn commented:
If there are no issues left, I think this is ready.

pv commented:
I'd suggest addin a for loop over supported dtypes, so that you also test
the float32 variant.

24.12.2016 12.05 "Ilhan Polat" <notifications@github.com> kirjoitti:

> If there are no issues left, I think this is ready.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/6775#issuecomment-269077961>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5tNZIKEZ5f-2pgAIyQETY7OVu05Qks5rLO5IgaJpZM4KuyMt>
> .
>

ilayn commented:
@pv Which test do you have in mind?

pv commented:
Enough to cover all the lapack routine variants (sy/po/he/ge).

I also note that the tests added here many contain hardcoded matrices. It probably is more concise to use random matrices instead, and test the solution satisfies `assert_allclose(a.dot(x) - b, 0, atol=...)`, as I guess the tests are probably mostly smoketests  rather than checking eg forward vs. backward error.

It also cam be preferable to avoid code duplication in the test by using for loops eg like
```
sizes = [...]
assume_as = ['gen', 'sym', 'pos', 'her']
dtypes = [np.float32, np.float64, np.complex64, np.complex128]
for size, assume_a, dtype in itertools.product(sizes, assume_as, dtypes):
    is_complex = dtype in (np.complex64, np.complex128)
    if assume_a == 'her' and not is_complex:
        continue

    err_msg = "{} {} {}".format(size, assume_a, dtype)

    a = np.random.randn(size, size).astype(dtype)
    b = np.random.randn(size).astype(dtype)
    if is_complex:
        a = a + (1j*np.random.randn(size, size)).astype(dtype)

    if assume_a in ('sym', 'her'):
        a = a + a.T.conj()
    elif assume_a == 'pos':
        a = a.dot(a)

    ...
    assert_allclose(a.dot(x), b, atol=1e-12, rtol=1e-12, err_msg=err_msg)
```
etc

ilayn commented:
Thanks @pv. Added those and a few trivial ones for codecov. 

Regarding the hardcoded tests, I've left them as is to see whether it is in accordance with the old version of solve (1D array shapes and empty array behavior etc.) If this goes through we can later reorganize the tests and make them more to the point and concise.

ilayn commented:
happy new year ping :)

pv commented on scipy/linalg/tests/test_basic.py:
```diff
@@ -602,6 +605,150 @@ def test_check_finite(self):
+            elif assume_a == 'pos':
+                a = a.conj().T.dot(a) + 0.1*np.eye(size)
+
+            x = solve(a, b)
```
No assume_a here.

ilayn replied:
My screw up with github probably. Locally I do have it. Will fix.

pv commented on scipy/linalg/tests/test_basic.py:
```diff
@@ -602,6 +605,150 @@ def test_check_finite(self):
+                a = a.conj().T.dot(a) + 0.1*np.eye(size)
+
+            x = solve(a, b)
+            tol_ = 1e-12 if dtype_ in (np.float64, np.complex128) else 1e-5
```
Why underscore?

pv commented:
Typically, this is only done when it is needed to avoid shadowing.
Otherwise, it is not standard practice in scipy.

2.1.2017 1.58 "Ilhan Polat" <notifications@github.com> kirjoitti:

> *@ilayn* commented on this pull request.
> ------------------------------
>
> In scipy/linalg/tests/test_basic.py
> <https://github.com/scipy/scipy/pull/6775>:
>
> > +
> +            err_msg = ("Failed for size: {}, assume_a: {},"
> +                       "dtype: {}".format(size, assume_a, dtype_))
> +
> +            a = np.random.randn(size, size).astype(dtype_)
> +            b = np.random.randn(size).astype(dtype_)
> +            if is_complex:
> +                a = a + (1j*np.random.randn(size, size)).astype(dtype_)
> +
> +            if assume_a in ('sym', 'her'):
> +                a = a + a.T.conj()
> +            elif assume_a == 'pos':
> +                a = a.conj().T.dot(a) + 0.1*np.eye(size)
> +
> +            x = solve(a, b)
> +            tol_ = 1e-12 if dtype_ in (np.float64, np.complex128) else 1e-5
>
> Just to distinguish variables from the signature keywords with standard
> python practice
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/6775>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5tjpQWkhpR18M1EzfMuqQjYPZ2Ntks5rOD2AgaJpZM4KuyMt>
> .
>

ilayn commented:
I think all comments are addressed. Let me know if anything else is needed.

pv commented on scipy/linalg/basic.py:
```diff
@@ -75,34 +101,116 @@ def solve(a, b, sym_pos=False, lower=False, overwrite_a=False,
-    if a1.shape[0] != b1.shape[0]:
-        raise ValueError('incompatible dimensions')
-    overwrite_a = overwrite_a or _datacopied(a1, a)
-    overwrite_b = overwrite_b or _datacopied(b1, b)
```
These appear to have been removed?

ilayn commented:
@pv I see your comments but singe merged I don't know how to address them. Should I open a new PR for those and the `debug` keyword removal discussion ?

pv commented:
Yes, you can open a new PR for those. The PR for the `debug` keyword should also be separate, so that each PR is focused only on a single topic.
